web/packages/extension/js/util.js

```diff
@@ -0,0 +1,200 @@
+    dict_equality,
+    tab_query,
+    tab_sendmessage,
+    add_storage_change_listener,
+    open_settings_page,
+    set_message_listener,
+    get_extension_url,
+};
+
+function get_i18n_string(key) {
+    if (chrome && chrome.i18n && chrome.i18n.getMessage) {
+        return chrome.i18n.getMessage(key);
+    } else if (browser && browser.i18n && browser.i18n.getMessage) {
+        return browser.i18n.getMessage(key);
+    } else {
+        console.error("Can't get i18n message: " + key);
+    }
+}
+
+function set_sync_storage(key) {
+    if (
+        chrome &&
+        chrome.storage &&
+        chrome.storage.sync &&
+        chrome.storage.sync.set
+    ) {
+        chrome.storage.sync.set(key);
+    } else if (
+        browser &&
+        browser.storage &&
+        browser.storage.sync &&
+        browser.storage.sync.set
+    ) {
+        browser.storage.sync.set(key);
+    } else {
+        console.error("Can't set settings.");
+    }
+}
+
+function get_sync_storage(key, callback) {
+    if (
+        chrome &&
+        chrome.storage &&
+        chrome.storage.sync &&
+        chrome.storage.sync.get
+    ) {
+        chrome.storage.sync.get(key, callback);
+    } else if (
+        browser &&
+        browser.storage &&
+        browser.storage.sync &&
+        browser.storage.sync.get
+    ) {
+        browser.storage.sync.get(key, callback);
+    } else {
+        console.error("Couldn't read setting: " + key);
+    }
+}
+
+function add_storage_change_listener(listener) {
+    if (
+        chrome &&
+        chrome.storage &&
+        chrome.storage.onChanged &&
+        chrome.storage.onChanged.addListener
+    ) {
+        chrome.storage.onChanged.addListener(listener);
+    } else if (
+        browser &&
+        browser.storage &&
+        browser.storage.onChanged &&
+        browser.storage.onChanged.addListener
+    ) {
+        browser.storage.onChanged.addListener(listener);
+    } else {
+        console.error("Couldn't add setting change listener");
+    }
+}
+
+function reload_tab(tab, callback) {
+    if (chrome && chrome.tabs && chrome.tabs.reload) {
+        chrome.tabs.reload(tab, callback);
+    } else if (browser && browser.tabs && browser.tabs.reload) {
+        browser.tabs.reload(tab, callback);
+    } else {
+        console.error("Couldn't reload tab.");
+    }
+}
+
+function dict_equality(dict1, dict2) {
+    let is_equal = true;
+
+    for (var k in dict1) {
+        if (Object.prototype.hasOwnProperty.call(dict1, k)) {
+            is_equal = is_equal && dict1[k] === dict2[k];
+        }
+    }
+
+    for (let k in dict2) {
+        if (Object.prototype.hasOwnProperty.call(dict2, k)) {
+            is_equal = is_equal && dict1[k] === dict2[k];
+        }
+    }
+
+    return is_equal;
+}
+
+/**
+ * Promise-based version of `chrome.tabs.query`.
+ *
+ * Mozilla does this by default in `browser.tabs` but Chrome is behind on this
+ * sort of thing. Chrome won't even let us check if we're running in
```

Comment:
Running in what?! The suspense is killing me :D

Reply:
I noticed this kind of thing a lot. They were all like that from @kmeisthax 's original branch.